apiserver/facades/client/usermanager/usermanager_test.go

```diff
@@ -821,3 +821,88 @@ func (s *userManagerSuite) TestRemoveUserBulkSharedModels(c *gc.C) {
 	c.Assert(alice.IsDeleted(), jc.IsTrue)
 
 }
+
+func (s *userManagerSuite) TestResetPassword(c *gc.C) {
+	alex := s.Factory.MakeUser(c, &factory.UserParams{Name: "alex", NoModelUser: true})
+	c.Assert(alex.PasswordValid("password"), jc.IsTrue)
+
+	args := params.Entities{Entities: []params.Entity{{Tag: alex.Tag().String()}}}
+	results, err := s.usermanager.ResetPassword(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 1)
+
+	err = alex.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results[0].Tag, gc.DeepEquals, alex.Tag().String())
+	c.Assert(results.Results[0].SecretKey, gc.DeepEquals, alex.SecretKey())
+	c.Assert(alex.PasswordValid("password"), jc.IsFalse)
+}
+
+func (s *userManagerSuite) TestBlockResetPassword(c *gc.C) {
+	alex := s.Factory.MakeUser(c, &factory.UserParams{Name: "alex", NoModelUser: true})
+	args := params.Entities{Entities: []params.Entity{{Tag: alex.Tag().String()}}}
+	c.Assert(alex.PasswordValid("password"), jc.IsTrue)
+
+	s.BlockAllChanges(c, "TestBlockResetPassword")
+	_, err := s.usermanager.ResetPassword(args)
+	// Check that the call is blocked
+	s.AssertBlocked(c, err, "TestBlockResetPassword")
+
+	err = alex.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(alex.PasswordValid("password"), jc.IsTrue)
+}
+
+func (s *userManagerSuite) TestResetPasswordForSelf(c *gc.C) {
+	alex, err := s.State.User(s.AdminUserTag(c))
+	c.Assert(err, jc.ErrorIsNil)
+	args := params.Entities{Entities: []params.Entity{{Tag: alex.Tag().String()}}}
+	c.Assert(alex.PasswordValid("dummy-secret"), jc.IsTrue)
+
+	results, err := s.usermanager.ResetPassword(args)
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results, gc.HasLen, 1)
+
+	err = alex.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(results.Results[0].Tag, gc.DeepEquals, alex.Tag().String())
+	c.Assert(results.Results[0].SecretKey, gc.DeepEquals, alex.SecretKey())
+	c.Assert(alex.PasswordValid("dummy-secret"), jc.IsFalse)
+}
+
+func (s *userManagerSuite) TestResetPasswordNotControllerAdmin(c *gc.C) {
+	alex := s.Factory.MakeUser(c, &factory.UserParams{Name: "alex", NoModelUser: true})
+	c.Assert(alex.PasswordValid("password"), jc.IsTrue)
+	usermanager, err := usermanager.NewUserManagerAPI(
+		s.State, s.resources, apiservertesting.FakeAuthorizer{Tag: alex.Tag()})
+	c.Assert(err, jc.ErrorIsNil)
+
+	args := params.Entities{Entities: []params.Entity{{Tag: alex.Tag().String()}}}
+	_, err = usermanager.ResetPassword(args)
+	c.Assert(err, gc.ErrorMatches, "permission denied")
+
+	err = alex.Refresh()
+	c.Assert(err, jc.ErrorIsNil)
+	c.Assert(alex.PasswordValid("password"), jc.IsTrue)
+}
+
+func (s *userManagerSuite) TestResetPasswordFail(c *gc.C) {
```

Comment:
Its more about "having one that cannot be done doesn't mean we won't do the ones that we *can* do".
Otherwise it may be that we short-circuit as soon as we see one value we can't change, unhandled exceptions, etc. Maybe a bit too much belt-and-suspenders, but the "what is the side effect on X when Y fails" is often something that is worth having an explicit case for.